src/logic/map.cc

```diff
@@ -540,154 +540,291 @@ void Map::set_origin(const Coords& new_origin) {
 	log("Map origin was shifted by (%d, %d)\n", new_origin.x, new_origin.y);
 }
 
-/* Helper function for resize():
- * Calculates the coords of 'c' after resizing the map from the given old size to the given new size
- * at 'split'.
- */
-static Coords transform_coords(const Coords& c,
-                               const Coords& split,
-                               int16_t w_new,
-                               int16_t h_new,
-                               int16_t w_old,
-                               int16_t h_old) {
-	const int16_t delta_w = w_new - w_old;
-	const int16_t delta_h = h_new - h_old;
-	if (c.x < split.x && c.y < split.y) {
-		// Nothing to shift
-		Coords result(c);
-		Map::normalize_coords(result, w_new, h_new);
-		return result;
-	} else if ((w_new < w_old && c.x >= split.x && c.x < split.x - delta_w) ||
-	           (h_new < h_old && c.y >= split.y && c.y < split.y - delta_h)) {
-		// Field removed
-		return Coords::null();
-	}
-	Coords result(c.x, c.y);
-	if (c.x >= split.x) {
-		result.x += delta_w;
-	}
-	if (c.y >= split.y) {
-		result.y += delta_h;
-	}
-	Map::normalize_coords(result, w_new, h_new);
-	return result;
-}
-
-/* Change the size of the (already initialized) map to 'w'×'h' by inserting/deleting fields south
- * and east of 'split'. Returns the data of fields that were deleted during resizing. This function
- * will notify all players of the change in map size, but not of anything else. This is because the
- * editor may want to do some post-resize cleanup first, and this function is intended to be used
- * only by the editor anyway. You should call recalc_whole_map() afterwards to resolve height
- * differences etc.
- */
-std::map<Coords, FieldData>
-Map::resize(EditorGameBase& egbase, const Coords split, const int32_t w, const int32_t h) {
-	assert(w > 0);
-	assert(h > 0);
-
-	std::map<Coords, FieldData> deleted;
+void Map::resize(EditorGameBase& egbase, const Coords split, const int32_t w, const int32_t h) {
 	if (w == width_ && h == height_) {
-		return deleted;
+		return;
 	}
 
-	const uint32_t field_size = w * h;
-	const uint32_t old_field_size = width_ * height_;
-
-	std::unique_ptr<Field[]> new_fields(new Field[field_size]);
-	clear_array<>(&new_fields, field_size);
-
-	// Take care of starting positions and port spaces
-	for (uint8_t i = get_nrplayers(); i > 0; --i) {
-		if (starting_pos_[i - 1]) {
-			starting_pos_[i - 1] =
-			   transform_coords(starting_pos_[i - 1], split, w, h, width_, height_);
+	const int16_t dx = w - width_;
+	const int16_t dy = h - height_;
+	const int16_t xoff = split.x - w;
+	const int16_t yoff = split.y - h;
+
+	// returns the old coords for the given new coords
+	std::function<Coords(int, int)> lambda_co;
+	// transform the given player starting pos or port space to new coordinate space
+	std::function<Coords(Coords)> lambda_cn;
+
```

Comment:
// coordinates may not be / will be normalized after mapping/projection ?
// Applying both transformation will result in the orginal coordinates unless a point was dropped
// or this is a bijective mapping as long as the coordinates from the smaller map are involed
-> You should make this an assertion jus in case some +/- error slipped through your fingers....